src/credentials/sso/clients.ts

```diff
@@ -0,0 +1,192 @@
+    (this: void, ...args: infer P3): infer R3
+}
+    ? (...args: P1) => R1
+    : never
+
+export class SsoClient {
+    public constructor(private readonly client: SSO, private readonly provider: SsoAccessTokenProvider) {}
+
+    public listAccounts(
+        request: Omit<ListAccountsRequest, OmittedProps> = {}
+    ): AsyncCollection<RequiredProps<AccountInfo, 'accountId'>[]> {
+        const method = this.client.listAccounts.bind(this.client)
+        const requester = (request: Omit<ListAccountsRequest, 'accessToken'>) =>
+            this.call(method as ExtractOverload<typeof method>, request)
+
+        const collection = pageableToCollection(requester, request, 'nextToken', 'accountList')
+
+        return collection.filter(isNonNullable).map(accounts => accounts.map(a => (assertHasProps(a, 'accountId'), a)))
+    }
+
+    public listAccountRoles(
+        request: Omit<ListAccountRolesRequest, OmittedProps>
+    ): AsyncCollection<Required<RoleInfo>[]> {
+        const method = this.client.listAccountRoles.bind(this.client)
+        const requester = (request: Omit<ListAccountRolesRequest, 'accessToken'>) =>
+            this.call(method as ExtractOverload<typeof method>, request)
+
+        const collection = pageableToCollection(requester, request, 'nextToken', 'roleList')
+
+        return collection
+            .filter(isNonNullable)
+            .map(roles => roles.map(r => (assertHasProps(r, 'roleName', 'accountId'), r)))
+    }
+
+    public async getRoleCredentials(request: Omit<GetRoleCredentialsRequest, OmittedProps>) {
+        const method = this.client.getRoleCredentials.bind(this.client)
+        const response = await this.call(method as ExtractOverload<typeof method>, request)
+
+        assertHasProps(response, 'roleCredentials')
+        assertHasProps(response.roleCredentials, 'accessKeyId', 'secretAccessKey')
+
+        const expiration = response.roleCredentials.expiration
+
+        return {
+            ...response.roleCredentials,
+            expiration: expiration ? new globals.clock.Date(expiration) : undefined,
+        }
+    }
+
+    private call<T extends { accessToken: string | undefined }, U>(
+        method: (request: T) => Promise<U>,
+        request: Omit<T, 'accessToken'>
+    ): Promise<U> {
+        const requester = async (req: T) => {
+            const token = await this.provider.getToken()
+            assertHasProps(token, 'accessToken')
+
+            try {
+                return await method({ ...req, accessToken: token.accessToken })
+            } catch (error) {
+                await this.handleError(error)
+                throw error
+            }
+        }
+
+        return requester(request as T)
+    }
+
+    private async handleError(error: unknown): Promise<never> {
+        if (
+            error instanceof SSOServiceException &&
+            error.$fault === 'client' &&
+            !(isThrottlingError(error) || isTransientError(error))
```

Comment:
interesting... should we mention `@aws-sdk/service-error-classification` somewhere in our guidelines, or maybe in ToolkitError class description?

Reply:
I'll put a util function in `errors.ts`, maybe `isClientFault`.